sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSchemaInferenceSuite.scala

```diff
@@ -0,0 +1,352 @@
+      StructField(
+        name = field,
+        dataType = LongType,
+        nullable = true,
+        metadata = new MetadataBuilder().putString(HIVE_TYPE_STRING, "bigint").build())
+    }
+    // and all partition columns as ints
+    val partitionStructFields = partitionCols.map { field =>
+      StructField(
+        // Partition column case isn't preserved
+        name = field.toLowerCase,
+        dataType = IntegerType,
+        nullable = true,
+        metadata = new MetadataBuilder().putString(HIVE_TYPE_STRING, "int").build())
+    }
+    val schema = StructType(structFields ++ partitionStructFields)
+
+    // Write some test data (partitioned if specified)
+    val writer = spark.range(NUM_RECORDS)
+      .selectExpr((fields ++ partitionCols).map("id as " + _): _*)
+      .write
+      .partitionBy(partitionCols: _*)
+      .mode("overwrite")
+    fileType match {
+      case ORC_FILE_TYPE =>
+       writer.orc(dir.getAbsolutePath)
+      case PARQUET_FILE_TYPE =>
+       writer.parquet(dir.getAbsolutePath)
+    }
+
+    // Create Hive external table with lowercased schema
+    val serde = HiveSerDe.serdeMap(fileType)
+    client.createTable(
+      hiveExternalCatalogTable(
+        tableName,
+        serde,
+        dir.getAbsolutePath,
+        lowerCaseSchema(schema),
+        partitionCols.map(_.toLowerCase)),
+      true)
+
+    // Add partition records (if specified)
+    if (!partitionCols.isEmpty) {
+      val partitions = (0 until NUM_RECORDS).map { idx =>
+         hiveCatalogPartition(dir.getAbsolutePath, serde, partitionCols, idx)
+      }.toSeq
+      client.createPartitions(DATABASE, tableName, partitions, true)
+    }
+
+    // Check that the table returned by HiveExternalCatalog has schemaPreservesCase set to false
+    // and that the raw table returned by the Hive client doesn't have any Spark SQL properties
+    // set (table needs to be obtained from client since HiveExternalCatalog filters these
+    // properties out).
+    assert(!externalCatalog.getTable(DATABASE, tableName).schemaPreservesCase)
+    val rawTable = client.getTable(DATABASE, tableName)
+    assert(rawTable.properties.filterKeys(_.startsWith(SPARK_SQL_PREFIX)) == Map.empty)
+    schema
+  }
+
+  private def withTestTables(
+    fileType: String)(f: (String, Seq[String], Seq[String], StructType) => Unit): Unit = {
+    // Test both a partitioned and unpartitioned Hive table
+    val tableFields = Seq(
+      (Seq("fieldOne"), Seq("partCol1", "partCol2")),
+      (Seq("fieldOne", "fieldTwo"), Seq.empty[String]))
+
+    tableFields.foreach { case (fields, partCols) =>
+      withTempDir { dir =>
+        val tableName = "test_table"
+        val schema = setupExternalTable(tableName, fileType, fields, partCols, dir)
+        try f(
+          tableName,
+          fields,
+          partCols,
+          schema) finally spark.sql(s"DROP TABLE IF EXISTS $tableName")
+      }
+    }
+  }
+
+  private def withFileTypes(f: (String) => Unit): Unit
```

Comment:
I didn't do this since I wanted to be able to wrap the  ```test("...") { ... }``` block in ```withFileTypes { ... }``` in order to have individual tests per file type. The thought here is that this will help identify if schema inference is broken just for a particular file format.